### ISSUE TITLE
Running an existing executable raises a ProcessException: No such file or directory

### ISSUE THREAD

**Issue by MaksimDan** [flutter/flutter#65983]
## Steps to Reproduce

1. Create an app: flutter create --org foobar baz
2. In your pubspec.yml add an executable to your assets. Mine just does an std_out of "hello world" 

```
flutter:
  uses-material-design: true
  assets:
    - lib/py/dist/main
```

3. In main.dart, write the executable on load, and execute it _incrementCounter()

```dart
import 'dart:io';

import 'package:flutter/material.dart';
import 'package:flutter/services.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);
  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  int _counter = 0;
  static String executablePath = '/data/data/foobar.baz/main';

  void writeExecutable() async {
    if (await File(executablePath).exists()) {
      File(executablePath).delete();
      print('deleted old executable');
    } else {
      print('not executable exists');
    }

    ByteData data = await rootBundle.load('lib/py/dist/main');
    List<int> bytes =
        data.buffer.asUint8List(data.offsetInBytes, data.lengthInBytes);
    await File(executablePath).writeAsBytes(bytes);
    print('wrote new executable');
  }

  void invokeExecutable() async {
    Process.run('chmod', ['744', executablePath]).then((ProcessResult results) {
      Process.run(executablePath, []).then((ProcessResult results) {
        print(results.stdout);
      });
    });
  }

  Future<void> _incrementCounter() async {
    setState(() {
      _counter++;
      invokeExecutable();
    });
  }

  @override
  void initState() {
    writeExecutable();
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            Text(
              'You have pushed the button this many times:',
            ),
            Text(
              '$_counter',
              style: Theme.of(context).textTheme.display1,
            ),
          ],
        ),
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: _incrementCounter,
        tooltip: 'Increment',
        child: Icon(Icons.add),
      ),
    );
  }
}

```

3. Run the application (in my case, android)
4. Press FloatingActionButton

**Expected results:** 

```
hello world
```

**Actual results:** 

```
E/flutter (23390): [ERROR:flutter/lib/ui/ui_dart_state.cc(166)] Unhandled Exception: ProcessException: No such file or directory
E/flutter (23390):   Command: /data/data/foobar.baz/main
```

I used the Device File Explore provided by android tools, and validated that executable exists at: `/data/data/foobar.baz/main`

<details>
  <summary>Logs</summary>

```
[  +12 ms] Connected to _flutterView/0xec2c5ac0.
[   +1 ms] Flutter run key commands.
[   +2 ms] r Hot reload. 🔥🔥🔥
[   +1 ms] R Hot restart.
[        ] h Repeat this help message.
[        ] d Detach (terminate "flutter run" but leave application running).
[        ] c Clear the screen
[        ] q Quit (terminate the application on the device).
[        ] An Observatory debugger and profiler on sdk gphone x86 is available at: http://127.0.0.1:50207/VK62ELYRT78=/
[ +690 ms] I/flutter (23557): deleted old executable
[ +251 ms] I/flutter (23557): wrote new executable
[+11855 ms] E/flutter (23557): [ERROR:flutter/lib/ui/ui_dart_state.cc(166)] Unhandled Exception: ProcessException: No such file or directory
[        ] E/flutter (23557):   Command: /data/data/com.maksimdan.face_merger/main 
[        ] E/flutter (23557): 
[+2257 ms] E/flutter (23557): [ERROR:fl
```

```
Analyzing face_merger...                                                

   info • 'display1' is deprecated and shouldn't be used. This is the term used in the 2014 version of material design. The modern term is headline4. This feature was deprecated after v1.13.8. • lib/main.dart:84:50 • deprecated_member_use

1 issue found. (ran in 1.9s)

```

```
face_merger % flutter doctor -v
[✓] Flutter (Channel stable, 1.20.3, on Mac OS X 10.15.6 19G2021, locale en-US)
    • Flutter version 1.20.3 at /Users/danielm/Documents/flutter
    • Framework revision 216dee60c0 (2 weeks ago), 2020-09-01 12:24:47 -0700
    • Engine revision d1bc06f032
    • Dart version 2.9.2

 
[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.1)
    • Android SDK at /Users/danielm/Library/Android/sdk
    • Platform android-30, build-tools 30.0.1
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)
    • All Android licenses accepted.

[!] Xcode - develop for iOS and macOS
    ✗ Xcode installation is incomplete; a full installation is necessary for iOS development.
      Download at: https://developer.apple.com/xcode/download/
      Or install Xcode via the App Store.
      Once installed, run:
        sudo xcode-select --switch /Applications/Xcode.app/Contents/Developer
        sudo xcodebuild -runFirstLaunch
    • CocoaPods version 1.9.3

[✓] Android Studio (version 4.0)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 47.1.2
    • Dart plugin version 193.7361
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)

[✓] IntelliJ IDEA Ultimate Edition (version 2020.1.3)
    • IntelliJ at /Applications/IntelliJ IDEA.app
    • Flutter plugin version 47.1.3
    • Dart plugin version 201.8538.43

[✓] VS Code (version 1.48.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.14.1

[✓] Connected device (1 available)
    • sdk gphone x86 (mobile) • emulator-5554 • android-x86 • Android 11 (API 30) (emulator)

! Doctor found issues in 1 category.
```

</details>

**Comment by pedromassangocode** [comment]
Hi @MaksimDan 
Your code for writing the file is wrong:

> ```dart
>     await File(executablePath).writeAsBytes(bytes);
> ```

In order to write a file you need write permissions on Android and you need to check if the full directory path already exist before trying to create the file, or you can create it if does not exist by:

```dart
    var file = File(executablePath);
    file = await file.create(recursive: true); // recursive means: create the full directory if it does not exist
    file.writeAsBytes(bytes);
```

Please learn more at https://flutter.dev/docs/cookbook/persistence/reading-writing-files

Please see https://flutter.dev/community for resources and asking questions like this,
you may also get some help if you post it on Stack Overflow and if you need help with your code, please see https://www.reddit.com/r/flutterhelp/
Closing, as this isn't an issue with Flutter itself. If you disagree, please write in the comments and I will reopen it.
Thank you

**Comment by MaksimDan** [comment]
Before going to flutter/issues on github, I made sure to ask this question on stackoverflow: https://stackoverflow.com/questions/62056764/hosting-an-executable-within-a-flutter-application. This was 3 months ago.

I've tried the solution you've suggest as well as read the documentation, but the issue persists.

In any case, between the exception thrown telling me that there is no file or directory, and me validating that the file does exist through the device file explore tells me that somewhere something is misinterpreted.

Thanks for your feedback
